src/blockchain_db/sqlite/db_sqlite.cpp

```diff
@@ -0,0 +1,666 @@
+      if (!success)
+        return false;
+    }
+
+    // Add back to the database payments that had been made in this block
+    delete_block_payments(block_height);
+
+    if (decrement_height())
+      success = true;
+    transaction.commit();
+  }
+  catch (std::exception& e)
+  {
+    MFATAL("Exception: " << e.what());
+    return false;
+  }
+  return success;
+}
+
+bool BlockchainSQLite::validate_batch_payment(std::vector<std::tuple<crypto::public_key, uint64_t>> miner_tx_vouts, std::vector<cryptonote::batch_sn_payment> calculated_payments_from_batching_db, uint64_t block_height)
+{
+  LOG_PRINT_L3("BlockchainDB_SQLITE::" << __func__);
+  size_t length_miner_tx_vouts = miner_tx_vouts.size();
+  size_t length_calculated_payments_from_batching_db = calculated_payments_from_batching_db.size();
+
+  if (length_miner_tx_vouts != length_calculated_payments_from_batching_db)
+  {
+    MERROR("Length of batch paments does not match, block vouts: " << length_miner_tx_vouts << " batch size: " << length_calculated_payments_from_batching_db);
+    return false;
+  }
+
+  int8_t vout_index = 0;
+  uint64_t total_oxen_payout_in_our_db = std::accumulate(calculated_payments_from_batching_db.begin(),calculated_payments_from_batching_db.end(), uint64_t(0), [](auto const a, auto const b){return a + b.amount;});
+  uint64_t total_oxen_payout_in_vouts = 0;
+  std::vector<batch_sn_payment> finalised_payments;
+  cryptonote::keypair const deterministic_keypair = cryptonote::get_deterministic_keypair_from_height(block_height);
+  for(auto & vout : miner_tx_vouts)
+  {
+    if (std::get<1>(vout) != calculated_payments_from_batching_db[vout_index].amount)
+    {
+      MERROR("Service node reward amount incorrect. Should be " << cryptonote::print_money(calculated_payments_from_batching_db[vout_index].amount) << ", is: " << cryptonote::print_money(std::get<1>(vout)));
+      return false;
+    }
+    crypto::public_key out_eph_public_key{};
+    if (!cryptonote::get_deterministic_output_key(calculated_payments_from_batching_db[vout_index].address_info.address, deterministic_keypair, vout_index, out_eph_public_key))
+    {
+      MERROR("Failed to generate output one-time public key");
+      return false;
+    }
+    if (tools::view_guts(std::get<0>(vout)) != tools::view_guts(out_eph_public_key))
+    {
+      MERROR("Output ephemeral public key does not match");
+      return false;
+    }
+    total_oxen_payout_in_vouts += std::get<1>(vout);
+    finalised_payments.emplace_back(calculated_payments_from_batching_db[vout_index].address, std::get<1>(vout), m_nettype);
+    vout_index++;
+  }
+  if (total_oxen_payout_in_vouts != total_oxen_payout_in_our_db)
+  {
+    MERROR("Total service node reward amount incorrect. Should be " << cryptonote::print_money(total_oxen_payout_in_our_db) << ", is: " << cryptonote::print_money(total_oxen_payout_in_vouts));
+    return false;
+  }
+
+  return save_payments(block_height, finalised_payments);
```

Comment:
should return `true` and wherever you call this should call `save_payments` itself

Reply:
Ideally only the `save` part exists in this db layer, and the validation and calculation logic is handled separately.